lib/spack/spack/fetch_strategy.py

```diff
@@ -283,6 +291,17 @@ def check(self):
                     "%s checksum failed for %s" % (checker.hash_name, self.archive_file),
                     "Expected %s but got %s" % (self.digest, checker.sum))
 
+
+    def search_cache(self):
```

Comment:
If you implement the caching so that it builds a mirror, this part can be hoisted up to `Package.fetch()`, so that the cache is the first mirror it checks.  I think you would only need to add the `var/spack/cache` dir to the mirror roots in `Package.fetch()` to get this to work.  That would get you source repo archiving for free.